docs: Multistage Build guide (#3659)
dagger/dagger: docs/current/sdk/go/guides/544174-multistage-build.md

---
slug: /544174/multistage-build

> ---

Reviewer: Add displayed_sidebar: current or else the page will not have a sidebar

> # Multistage Builds

Reviewer: ```suggestion
# Use Dagger with Multi-stage Container Builds
```

> A common practice when building containers with Dockerfiles is something called a multistage build. This means within the `docker build`, your application is compiled in a context which has tools that are required for building your app, but not necessarily required for running the application. To reduce the number of dependencies in the image you actually run the application on, the compiled application is copied to a different base image which only has the required components to run the application. More information about multistage builds can be found in [this guide](https://docs.docker.com/build/building/multi-stage/) from Docker.

Reviewer: ```suggestion
Multi-stage builds are a common practice when building containers with Docker. 

- First, your application is compiled in a context which has tools that are required for building the application, but not necessarily required for running it. 
- Next, to reduce the number of dependencies and hence the size of the image, the compiled application is copied to a different base image which only has the required components to run the application. 

[Learn more about multi-stage builds in the Docker documentation](https://docs.docker.com/build/building/multi-stage/).
```

> A common practice when building containers with Dockerfiles is something called a multistage build. This means within the `docker build`, your application is compiled in a context which has tools that are required for building your app, but not necessarily required for running the application. To reduce the number of dependencies in the image you actually run the application on, the compiled application is copied to a different base image which only has the required components to run the application. More information about multistage builds can be found in [this guide](https://docs.docker.com/build/building/multi-stage/) from Docker.

Reviewer: Why not move this paragraph under Introduction?

> This document shows how multistage builds are created using the Go SDK.

Reviewer: ```suggestion
This guide explains how to perform multi-stage builds with the Go SDK.
```

## Introduction

## Requirements

This guide assumes that:

- You have a Go development environment with Go 1.15 or later. If not, [download and install Go](https://go.dev/doc/install).
- You are familiar with the basics of the Go SDK and have it installed. If not, read the [Go SDK guide](../959738-get-started.md) and the [Go SDK installation instructions](../371491-install.md).
- You have Docker installed and running on the host system. If not, [install Docker](https://docs.docker.com/engine/install/).

## Example

> The following code snippet demonstrates a multistage build with the Go SDK. Below the snippet, we'll break down exactly what's happening

Reviewer: s/multistage/multi-stage/g

> The following code snippet demonstrates a multistage build with the Go SDK. Below the snippet, we'll break down exactly what's happening

Reviewer: ```suggestion
The following code snippet demonstrates a multistage build with the Go SDK. 
```

```go file=../snippets/multistage-build/main.go

```

> The beginning of the file starts by creating a Dagger client and loading the project which is going to be built. Once we have a reference to the project, we build the application by using the `golang:latest` image to mount the source directory, set `CGO_ENABLED=` since the binary will be published on Alpine, and then execute `go build`.

Reviewer: ```suggestion
This code listing starts by creating a Dagger client and loading the project to be built. It obtains a reference to the project and then builds the application by using the `golang:latest` image to mount the source directory, sets `CGO_ENABLED=` since the binary will be published on `alpine`, and executes `go build`.
```

> Next, in the highlighted section, the multistage build is achieved by taking the build artifact from the build stage and putting it in an Alpine image.

Reviewer: ```suggestion
Next, in the highlighted section, the multi-stage build is achieved by transferring the build artifact from the builder image to a runtime image based on `alpine`. The steps are:
```

> - Create a new container which will be used as the runtime image. This is using `From("alpine")`

Reviewer: ```suggestion
- Create a new container image which will be used as the runtime image, using `From("alpine")`.
```

> - Include the build artifact from the builder image in the new container by replacing the container's filesystem with the original filesystem plus the build artifact

Reviewer: ```suggestion
- Transfer the build artifact from the builder image to the new container image by replacing the container image's filesystem with the original filesystem plus the build artifact.
```

> - Set the entrypoint to our application so that the application is executed by default when the container is run

Reviewer: ```suggestion
- Set the container entrypoint to the application so that it is executed by default when the container runs.
```

> Now that we have a slimmed down production image, it can be pushed to a registry and deployed!

Reviewer: ```suggestion
The final optimized image can now be pushed to a registry and deployed!
```
